src/core/lib/security/credentials/external/url_external_account_credentials.cc

```diff
@@ -121,7 +123,18 @@ void UrlExternalAccountCredentials::RetrieveSubjectToken(
   grpc_httpcli_request request;
   memset(&request, 0, sizeof(grpc_httpcli_request));
   request.host = const_cast<char*>(url_.authority().c_str());
-  request.http.path = gpr_strdup(url_.path().c_str());
+  if (url_.query_parameter_pairs().empty()) {
+    request.http.path = gpr_strdup(url_.path().c_str());
+  } else {
+    std::vector<std::string> query_vector;
+    for (const URI::QueryParam& query_kv : url_.query_parameter_pairs()) {
+      query_vector.emplace_back(
+          absl::StrCat(query_kv.key, "=", query_kv.value));
```

Comment:
This will suffer from percent-encoding problems. See for example: https://github.com/grpc/grpc/blob/master/test/core/uri/uri_parser_test.cc#L151-L153. A percent-decoded query value, which is what the URI class provides, can legitimately contain the character '&'. Without re-encoding it to "%26", the resulting path here won't be correct.

I'm not sure how many places we're reconstructing a URI string piece by piece, but it may be worth having a more generic way to re-encode URI components.

Reply:
Hmm, I think you're right. We actually need the original url string. So instead of decoding then encoding, I think I'll just store the original url here. Thanks for pointing this out!